src/rules/selector-max-depth/index.js

```diff
@@ -0,0 +1,78 @@
+import resolvedNestedSelector from "postcss-resolve-nested-selector"
+
+import {
+  isStandardRule,
+  isStandardSelector,
+  report,
+  ruleMessages,
+  validateOptions,
+} from "../../utils"
+
+export const ruleName = "selector-max-depth"
+
+export const messages = ruleMessages(ruleName, {
+  expected: (selector, depth) => `Expected "${selector}" to have a depth equal to or less than ${depth}`,
+})
+
+export default function (max) {
+  return (root, result) => {
+    const innerSelectorReg = /(?::not\((.*?)\))/g
+    const validOptions = validateOptions(result, ruleName, {
+      actual: max,
+      possible: [function (max) {
+        return typeof max === "number" && max > 0
+      }],
+    })
+    if (!validOptions) { return }
+
+    // Compares depth of a selector with the maximum
+    function checkDepth(selector, rule) {
```

Comment:
> So shall I create a separate issue with this question, or you guys would like to discuss it between yourselves?

We can discuss it here. We'll wait for @davidtheclark to weigh into the discussion and then sort it out after that.

> I really don't want to go Sheldon Cooper here forcing everyone my way, just want to make things clear since I see a serious defference in performance.

Your attention to performance detail is much appreciated :) Performance is definitely a key consideration, but we need to weigh up the other considerations (mainly the maintenance overhead going forward). Balancing needs and what not.